zerver/lib/actions.py

```diff
@@ -5144,6 +5144,14 @@ def do_change_stream_permission(
             )
 
         if old_is_web_public_value != stream.is_web_public:
+            # If `is_web_public` property of the stream was changed,
+            # we update all the attachments to reflect so.
+            # This avoids us expensive queries to the database
+            # when a spectator is trying to access these attachments.
+            Attachment.objects.filter(messages__recipient__id=stream.recipient.id).update(
+                is_web_public=stream.is_web_public
+            )
```

Comment:
This change is not correct in the presence of messages that might appear in more than one stream.

We can do the `=>True` change here safely, but for changing to `False`, it's not clear there's an efficient algorithm to do that.

I suppose we likely have the same bug with the `is_realm_public` flag on Attachment, too.

Reply:
I've added a commit to address this in a way that I believe is correct, including a test, by converting `is_web_public` to be a cache.